### PR TITLE
[Debugger] Allow to step into Xamarin code.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Adds the Xamarin.iOS and Xamarin.Mac sources, stops checkins in the install sources the path to $(TOP)/src in our tree and does a rebase to the locations for the sources (/Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/[Xamarin.iOS|Xamarin.Mac])

### CODE REVIEW

**Comment by mandel-macaque** [comment]
With the updated version the src dir looks similar to the following gist: https://gist.github.com/mandel-macaque/ee0543973586e91da6d9b5eb8f543cf2

**Comment by mandel-macaque** [comment]
There are some details to understand:

1. All xamarin sources, both manual and generated are under the framework dir so that we can step into the generated and non generated code.
2. The OpenTK sources are added.
3. The rutime delegates sources are not yet added, should they?
4. MDB rebase needs some changes to get both the generated and non generated paths correct.

**Comment by mandel-macaque** [comment]
Hash 9eebd986c9554b935bb5e48a5de64509e5ba415e adds the following in the src folders:

iOS -> https://gist.github.com/mandel-macaque/72ef8170b4612b00b54c51ee8b21c5a1
Mac -> https://gist.github.com/mandel-macaque/e4174abf675733421170998e55cf348c

File: builds/Makefile [hunk 1]
```diff
@@ -259,6 +259,8 @@ $$(MAC_DESTDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/$(1)/%.mdb: $$(BUILD_DESTD
 	$$(Q) install -m 0644 $$< $$@
 	$$(Q_MDB) $$(MDB_REBASE) -q -i $$(abspath $$(MONO_PATH))/ -o $$(MAC_TARGETDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/src/mono/ $$@
 	$$(Q) chmod 0644 $$@
+
```
Comment (rolfbjarne) [inline builds/Makefile]: Whitespace change.

File: tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (spouliot) [inline tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj]: any reason for a new .dll ? i.e. can't that just be compiled inside the existing tool ?

File: tools/install-source/Program.cs [hunk 3]
```diff
@@ -43,27 +58,26 @@ public static void Main (string[] arguments)
...
+			InstallDir = installDir,
+			MonoSourcePath = monopath,
+			XamarinSourcePath = xamarinpath,
+			FrameworkPath = (installDir.Contains ("Xamarin.iOS.framework")) ? "Xamarin.iOS.framework" : "Xamarin.Mac.framework",
+			OpenTKSourcePath = opentkpath,
+		};
 
+		Console.WriteLine ($"Install dir is {installDir}");
 		foreach (string mdb_file in mdb_files) {
+			Console.WriteLine("Mdb file is {0}", mdb_file);
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Please make debug output only show if verbose (or not at all :smile:)

File: tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				var relativePath = path.Remove (0, pos);
+				return Path.Combine (InstallDir, "src", FrameworkPath.Remove (FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture)), relativePath);	
+			}
+			return null;
+		}
+
+		public string GetTargetPath (string path)
+		{
+			if (IsRunrime (path))
+				return null; // TODO: Do we add this guys?
```
Comment (chamons) [inline tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs]: Please consider converting TODOs into bugzilla bugs if they still open when you land. TODO in code get forgotten.

File: tools/install-source/Program.cs (hunk 3 above)
Comment (chamons) [inline tools/install-source/Program.cs]: You can even do the trick that mmp/mtouch does and make verbose a number and have some output that only shows up with multiple -v's

File: tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj (hunk 2 above)
Comment (mandel-macaque) [inline tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj]: Well, I added it to be able tot do a nice unit testing of the diff path manglers. There is some logic between generated code and note. I can move it to the tool and see if I can reference it in the unit tests.

File: tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		{
+			return path.Contains (NativeTypeSubpath);
+		}
+
+		/// <summary>
+		/// Returns if the pat is from the runtime generated code.
+		/// </summary>
+		/// <returns><c>true</c>, if runrime was ised, <c>false</c> otherwise.</returns>
+		/// <param name="path">Path.</param>
+		static bool IsRunrime (string path)
```
Comment (mandel-macaque) [inline tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs]: Super bug.

File: tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs (hunk 4 above)
Comment (mandel-macaque) [inline tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs]: Yeah, it was a TODO for the reviewers to spot and complain ;)

File: tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs (hunk 5 above)
Comment (mandel-macaque) [inline tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs]: A meant typo ;)

File: tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj (hunk 2 above)
Comment (spouliot) [inline tools/install-source/InstallSourcesLib/InstallSourcesLib.csproj]: you should be able to reference the .exe to run the unit tests

File: tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+			else // we are dealing with a correct path (this does happen on the mac side)
+				src = src.Remove (0, src.IndexOf ("/src/", StringComparison.InvariantCulture) + "/src/".Length);
+			src = Path.Combine (XamarinSourcePath, src);
+			return src;
+		}
+
+		public string GetSourcePath(string path)
+		{
+			if (IsRunrime (path))
+				return null; // TODO: Do we add this guys?
```
Comment (spouliot) [inline tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs]: another TODO that should be fixed

File: src/Makefile [hunk 7]
```diff
@@ -618,6 +618,7 @@ $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/XamMac.dll: $(MAC_BUILD_DIR)
 $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/reference/mobile/Xamarin.Mac.dll: $(MAC_BUILD_DIR)/mobile-reference/Xamarin.Mac.dll | $(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/reference/mobile
 	$(Q) install -m 0755 $< $@
 	$(Q) install -m 0644 $<.mdb $@.mdb
+	$(Q) chmod 0644 $@.mdb
```
Comment (rolfbjarne) [inline src/Makefile]: This `chmod` is unnecessary, the previous `install` already does change the permissions.

Comment (mandel-macaque) [inline src/Makefile]: True. I also agree with the comments, I can wait for pr #1872 to land, then rework the rebase logic. @spouliot comments?

Comment (spouliot) [inline src/Makefile]: Not much harm in postponing - this is not a regression, nor a commonly requested feature.

**Comment by spouliot** [comment]
@mandel-macaque the mono-2017-02 PR was merged in master

File: builds/Makefile [hunk 8]
```diff
@@ -257,6 +257,12 @@ MAC_DIRECTORIES += \
 
 $$(MAC_DESTDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/$(1)/%: $$(BUILD_DESTDIR)/bcl/$(2)/% | $$(MAC_DESTDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/$(1)/Facades
 	$$(Q) install -m 0755 $$< $$@
+
+
+$$(MAC_DESTDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/$(1)/%.pdb: $$(BUILD_DESTDIR)/mac/$(2)/bcl/%.pdb | $$(MAC_DESTDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/$(1)/Facades
+	$$(Q) install -m 0644 $$< $$@
+	$$(Q_MDB) $$(MDB_REBASE) -q -i $$(abspath $$(MONO_PATH))/ -o $$(MAC_TARGETDIR)$$(MAC_FRAMEWORK_CURRENT_DIR)/src/mono/ $$@
```
Comment (rolfbjarne) [inline builds/Makefile]: You're calling MDB_REBASE on pdb, that won't work.

File: src/Makefile (hunk 7 above)
Comment (rolfbjarne) [inline src/Makefile]: The `chmod` is still unnecessary :smile:

File: tools/install-source/IPathMangler.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	/// Represents a class that knows how to change paths to ensure that the correct source path is used.
+	/// </summary>
+	public interface IPathMangler
+	{
+		/// <summary>
+		/// Returns the real source path for the given path found in an mdb.
+		/// </summary>
+		/// <returns>The source path related to the mdb path.</returns>
+		/// <param name="path">The path found in the mdb file.</param>
+		string GetSourcePath(string path);
```
Comment (rolfbjarne) [inline tools/install-source/IPathMangler.cs]: Style: space before opening parenthesis.

File: tools/install-source/IPathMangler.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+		/// <summary>
+		/// Returns the target were a source path should be installed.
+		/// </summary>
+		/// <returns>The target path for the installation.</returns>
+		/// <param name="path">The source path to install.</param>/
+		string GetTargetPath(string path);
```
Comment (rolfbjarne) [inline tools/install-source/IPathMangler.cs]: Style: space before opening parenthesis.

File: tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+{
+	[TestFixture]
+	public class MonoPathManglerTest
+	{
+		MonoPathMangler mangler;
+		string monoPath;
+		string installDir;
+
+		[SetUp]
+		public void SetUp()
```
Comment (rolfbjarne) [inline tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs]: Style: space before opening parenthesis.

File: tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+		{
+			monoPath = "/Users/test/xamarin-macios/external/mono/";
```
Comment (rolfbjarne) [inline tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs]: These paths look hard-coded to your local machine.

File: tools/install-source/InstallSourcesTests/XamarinSourcesPathManglerTest.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	[TestFixture]
+	public class XamarinSourcesPathManglerTest
+	{
+		string frameworkPath;
+		string xamarinSourcePath;
+		string installDir;
+		XamarinSourcesPathMangler mangler;
+
+		[SetUp]
+		public void SetUp()
```
Comment (rolfbjarne) [inline tools/install-source/InstallSourcesTests/XamarinSourcesPathManglerTest.cs]: Style: space before opening parenthesis.

File: tools/install-source/Makefile [hunk 14]
```diff
@@ -8,23 +9,65 @@ ifdef INCLUDE_MAC
...
+	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/mono/4.5/Xamarin.Mac.dll.mdb
+
+IOS_MDB_FILES = \
+	$(IOS_DESTDIR)$(MONOTOUCH_PREFIX)/lib/64bits/Xamarin.iOS.dll \
+	$(IOS_DESTDIR)$(MONOTOUCH_PREFIX)/lib/32bits/Xamarin.iOS.dll
+
+MAC_MDB_FILES = \
+	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/x86_64/mobile/Xamarin.Mac.dll.mdb \
+	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/i386/mobile/Xamarin.Mac.dll.mbd \
+	$(MAC_DESTDIR)$(MAC_FRAMEWORK_CURRENT_DIR)/lib/reference/full/Xamarin.Mac.dll.mdb \
```
Comment (rolfbjarne) [inline tools/install-source/Makefile]: You have a mix of `mdb` and `mbd` here.

File: tools/install-source/MonoPathMangler.cs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+		/// </summary>
+		/// <value>The install dir.</value>
+		public string InstallDir { get; set; }
+
+		public string GetSourcePath (string path) => path;
+
+		public string GetTargetPath (string path)
+		{
+			var relativePath = path.Substring (MonoSourcePath.Length);
+			if (relativePath.StartsWith ("/", StringComparison.InvariantCulture))
```
Comment (rolfbjarne) [inline tools/install-source/MonoPathMangler.cs]: This can use `StringComparison.Ordinal` (it's faster).

File: tools/install-source/OpenTKSourceMangler.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+		/// </summary>
+		/// <value>The install dir.</value>
+		public string InstallDir { get; set;}
+
+		public string GetSourcePath (string path) => path;
+
+		public string GetTargetPath (string path)
+		{
+			var relativePath = path.Substring (OpenTKSourcePath.Length);
+			if (relativePath.StartsWith ("/", StringComparison.InvariantCulture))
```
Comment (rolfbjarne) [inline tools/install-source/OpenTKSourceMangler.cs]: This can use StringComparison.Ordinal.

File: tools/install-source/Program.cs [hunk 17]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
 			return bool.Parse (value);
 		}
 	}
-	public static void Main (string[] arguments)
+
+	public static string FixPathEnding(string path)
+	{
+		if (!path.EndsWith(Path.DirectorySeparatorChar.ToString(), StringComparison.Ordinal))
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis (twice).

File: tools/install-source/Program.cs [hunk 18]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+			return path + Path.DirectorySeparatorChar;
+		return path;
+	}
+
+	// returns a tuple with the pdb files as the first value and the mdb files as the second
+	public static Tuple <HashSet<String>, HashSet<String>> GetSplittedPaths (List<string> paths)
+	{
+		var splittedPaths = new Tuple<HashSet<String>, HashSet<String>> (new HashSet<string> (), new HashSet<string> ());
+		foreach (var p in paths) {
+			if (p.EndsWith(".pdb", StringComparison.Ordinal)) {
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis.

File: tools/install-source/Program.cs [hunk 19]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+				splittedPaths.Item1.Add (p);
+				continue;
+			}
+			if (p.EndsWith(".mdb", StringComparison.Ordinal)) {
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis.

File: tools/install-source/Program.cs [hunk 20]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+	{
+		var srcs = new HashSet<String> ();
+		MonoSymbolFile symfile;
+
+		try
+		{
+			symfile = MonoSymbolFile.ReadSymbolFile(mdb_file);
+			srcs.UnionWith (from src in symfile.Sources select src.FileName);
+
+		}
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: `catch` clause on same line as the previous closing brace.

File: tools/install-source/Program.cs [hunk 21]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+	}
+
+	// returns the source paths used to create an assembly that has an mdb file
+	public static HashSet<String> GetFilePathsFromMdb (string mdb_file)
+	{
+		var srcs = new HashSet<String> ();
+		MonoSymbolFile symfile;
+
+		try
+		{
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: opening brace on the same line as the `try` clause.

File: tools/install-source/Program.cs [hunk 22]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+			symfile = MonoSymbolFile.ReadSymbolFile(mdb_file);
+			srcs.UnionWith (from src in symfile.Sources select src.FileName);
+
+		}
+		catch (IOException ioe)
+		{
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: opening brace on the same line as the `catch` clause.

File: tools/install-source/Program.cs [hunk 23]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+			Console.WriteLine("IO error while reading msb file '{0}': {1}", mdb_file, ioe.Message);
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis.

File: tools/install-source/Program.cs [hunk 24]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+
+	// returns the source paths used to create an assembly that has an mdb file
+	public static HashSet<String> GetFilePathsFromMdb (string mdb_file)
+	{
+		var srcs = new HashSet<String> ();
+		MonoSymbolFile symfile;
+
+		try
+		{
+			symfile = MonoSymbolFile.ReadSymbolFile(mdb_file);
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis.

File: tools/install-source/Program.cs [hunk 25]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
+		var writerParameters = new WriterParameters ();
+
+		var symbolReaderProvider = new PortablePdbReaderProvider ();
+		readerParameters.SymbolReaderProvider = symbolReaderProvider;
+		readerParameters.ReadSymbols = true;
+
+		var assemblyDefinition = AssemblyDefinition.ReadAssembly (assemblyPath, readerParameters);
+		var mainModule = assemblyDefinition.MainModule;
+		foreach (var type in mainModule.Types)
+		{
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: opening brace on the same line as the `foreach`.

File: tools/install-source/Program.cs [hunk 26]
```diff
@@ -25,87 +33,204 @@ static bool ParseBool (string value)
...
-				if (!src.StartsWith (monopath, StringComparison.Ordinal)) {
-					if (verbose)
-						Console.WriteLine ("{0}: not a mono source file", src);
-					continue;
-				}
-
-				srcs.Add (src);
-			}
+			var assemblySrcs = GetFilePathsFromMdb (mdbFile);
+			srcs.UnionWith(assemblySrcs);
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: Style: space before opening parenthesis.

File: tools/install-source/XamarinSourcesPathMangler.cs [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+		}
+
+		/// <summary>
+		/// Returns the source path for a generated file.
+		/// </summary>
+		/// <returns>The source path for native type.</returns>
+		/// <param name="path">Path.</param>
+		string GetSourcePathForGeneratedPath (string path)
+		{
+			var frameworkPrefix = FrameworkPath.Remove (FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture));
```
Comment (rolfbjarne) [inline tools/install-source/XamarinSourcesPathMangler.cs]: This can use `StringComparison.Ordinal`.

File: tools/install-source/XamarinSourcesPathMangler.cs [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+
+		/// <summary>
+		/// Returns the source path for the bindings that were manually done.
+		/// </summary>
+		/// <returns>The source path for manual source.</returns>
+		/// <param name="path">Path.</param>
+		string GetSourcePathForManualSource(string path)
+		{
+			var removalPath = $"/{FrameworkPath}/";
+			var srcFolder = Path.Combine ("src", FrameworkPath.Remove(FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture))) + Path.DirectorySeparatorChar;
```
Comment (rolfbjarne) [inline tools/install-source/XamarinSourcesPathMangler.cs]: Style: space before opening parenthesis.

This can use `StringComparison.Ordinal`.

File: tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs (hunk 12 above)
Comment (mandel-macaque) [inline tools/install-source/InstallSourcesTests/MonoPathManglerTest.cs]: Those are test paths, that is, we are just testing with a random user "test" and this path is passed to the tool in the makefile using the ones defined by make

File: tools/install-source/Program.cs [hunk 29]
```diff
@@ -25,87 +33,200 @@ static bool ParseBool (string value)
...
+		}
+
+		return srcs;
+	}
+
+	// returns the source paths used to create an assembly that has a pdb file
+	public static HashSet<String> GetFilePathsFromPdb (string pdbFile)
+	{
+		var assemblyPath = pdbFile.Replace (".pdb", ".dll");
+		var pdb = assemblyPath + ".pdb";
```
Comment (rolfbjarne) [inline tools/install-source/Program.cs]: This looks wrong, if you have `foo.dll`, the pdb is `foo.pdb`, not `foo.dll.pdb` (this is a change from mdb, where `foo.dll.mdb` existed).

File: tools/install-source/Makefile [hunk 30]
```diff
@@ -8,23 +9,65 @@ ifdef INCLUDE_MAC
 install-local:: install-source-mac
 endif
 
+MDB_REBASE_EXE = $(abspath $(MONO_PATH))mcs/class/lib/xammac/mdbrebase.exe
+MDB_REBASE = MONO_PATH=$(dir $(MDB_REBASE_EXE)) $(TOP)/builds/install/tools64/bin/mono --debug $(MDB_REBASE_EXE)
```
Comment (rolfbjarne) [inline tools/install-source/Makefile]: It doesn't look like you're using these two `MDB_REBASE` Make variables anywhere, in which case they can be removed.

File: tools/install-source/Makefile [hunk 31]
```diff
@@ -8,23 +9,65 @@ ifdef INCLUDE_MAC
...
+
+CECIL_PATH=$(MONO_PATH)/external/cecil
+CECIL_CONFIG=net_4_0_Release
+MONO_CECIL_DLL=$(CECIL_PATH)/bin/$(CECIL_CONFIG)/Mono.Cecil.dll
+MONO_CECIL_FILES=$(wildcard $(CECIL_PATH)/*.cs) $(wildcard $(CECIL_PATH)/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*/*/*.cs)
+MONO_CECIL_PDB_FILES=$(wildcard $(CECIL_PATH)/symbols/pdb/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*/*/*.cs)
```
Comment (rolfbjarne) [inline tools/install-source/Makefile]: `MONO_CECIL_PDB_FILES` isn't used anywhere either.

File: tools/install-source/Makefile [hunk 32]
```diff
@@ -8,23 +9,65 @@ ifdef INCLUDE_MAC
...
+
+Mono.Cecil.Pdb.dll: $(MONO_CECIL_PDB_DLL)
```
Comment (rolfbjarne) [inline tools/install-source/Makefile]: And `MONO_CECIL_PDB_DLL` isn't defined :smile:

**Comment by spouliot** [comment]
@mandel-macaque whats the status of the pr ? What are the failures ?

**Comment by mandel-macaque** [comment]
We have issues building the package in wrench, looking into those:

https://bugzilla.xamarin.com/attachment.cgi?id=24742

**Comment by mandel-macaque** [comment]
Latest push creates the src dir as follows => https://gist.github.com/mandel-macaque/57ed0b6d3142ea8b0b42caf456cabdb8

**Comment by ivanherreragl** [comment]
Verified. Ready to merge.

https://trello.com/c/Ee0FB9Ua/16-request-for-qa-test-step-into-framework-src

**Comment by rolfbjarne** [comment]
Is this expected?

https://gist.github.com/mandel-macaque/57ed0b6d3142ea8b0b42caf456cabdb8#file-src_example-txt-L4906-L4910

File: tools/install-source/Program.cs [hunk 33]
```diff
@@ -25,87 +33,233 @@ static bool ParseBool (string value)
...
+	}
+
+	// returns the source paths used to create an assembly that has a pdb file
+	public static HashSet<String> GetFilePathsFromPdb (string pdbFile)
+	{
+		var pdb = pdbFile.Replace (".pdb", ".dll");
+
+		var result = new HashSet<String> (StringComparer.OrdinalIgnoreCase);
+
+		if (!File.Exists (pdbFile) || !File.Exists (pdbFile))
```
Comment (marek-safar) [inline tools/install-source/Program.cs]: One is not enough?

File: tools/install-source/Program.cs [hunk 34]
```diff
@@ -25,87 +33,233 @@ static bool ParseBool (string value)
...
+			Console.WriteLine("IO error while reading msb file '{0}': {1}", mdb_file, ioe.Message);
+		}
+
+		return srcs;
+	}
+
+	// returns the source paths used to create an assembly that has a pdb file
+	public static HashSet<String> GetFilePathsFromPdb (string pdbFile)
+	{
+		var pdb = pdbFile.Replace (".pdb", ".dll");
```
Comment (marek-safar) [inline tools/install-source/Program.cs]: Path.ChangeExtension

File: tools/install-source/Makefile [hunk 35]
```diff
@@ -8,23 +9,62 @@ ifdef INCLUDE_MAC
...
 
+CECIL_PATH=$(MONO_PATH)/external/cecil
+CECIL_CONFIG=net_4_0_Release
+MONO_CECIL_DLL=$(CECIL_PATH)/bin/$(CECIL_CONFIG)/Mono.Cecil.dll
+MONO_CECIL_FILES=$(wildcard $(CECIL_PATH)/*.cs) $(wildcard $(CECIL_PATH)/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*/*.cs) $(wildcard $(CECIL_PATH)/*/*/*/*/*.cs)
+MONO_CECIL_PDB_FILES=$(wildcard $(CECIL_PATH)/symbols/pdb/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*/*.cs) $(wildcard $(CECIL_PATH)/symbols/pdb/*/*/*/*/*.cs)
+
+Mono.Cecil.Pdb.dll: $(MONO_CECIL_PDB_FILES)
+	$(Q_XBUILD) cd $(CECIL_PATH)/symbols/pdb/ && $(SYSTEM_XBUILD) /p:Configuration=$(CECIL_CONFIG) Mono.Cecil.Pdb.csproj $(XBUILD_VERBOSITY)
+	$(Q) cp $(CECIL_PATH)/symbols/pdb/obj/$(CECIL_CONFIG)/*.dll $(dir $@)
```
Comment (marek-safar) [inline tools/install-source/Makefile]: You should not copy from `obj` dir

File: tools/install-source/Program.cs (hunk 33 above)
Comment (mandel-macaque) [reply tools/install-source/Program.cs]: Good catch, I wonder what I was thinking there (probably copy paste in a refactoring).

**Comment by mandel-macaque** [comment]
@rolfbjarne you can see that now those paths are correct -> https://gist.github.com/mandel-macaque/57ed0b6d3142ea8b0b42caf456cabdb8

@spouliot paths that are /mcs/mcs/ have been ignored, removing the mcs code BUT not the mcs/class dir.

**Comment by mandel-macaque** [comment]
@ivanherreragl can you please take a second look?

File: tools/install-source/OpenTKSourceMangler.cs [hunk 36]
```diff
@@ -20,14 +23,27 @@ public class OpenTKSourceMangler : IPathMangler
 		/// <value>The install dir.</value>
 		public string InstallDir { get; set;}
 
-		public string GetSourcePath (string path) => path;
+		public string GetSourcePath (string path)
+		{
+			bool iosFramework = true;
+			if (path.StartsWith(OpenTKSourcePath, StringComparison.Ordinal))
```
Comment (rolfbjarne) [inline tools/install-source/OpenTKSourceMangler.cs]: Coding style: space before opening parenthesis.

File: tools/install-source/OpenTKSourceMangler.cs [hunk 37]
```diff
@@ -20,14 +23,27 @@ public class OpenTKSourceMangler : IPathMangler
...
+			return Path.Combine (OpenTKSourcePath, path);
+		}
 
 		public string GetTargetPath (string path)
 		{
 			var relativePath = path.Substring (OpenTKSourcePath.Length);
 			if (relativePath.StartsWith ("/", StringComparison.Ordinal))
 				relativePath = relativePath.Remove (0, 1);
-			var target = Path.Combine (InstallDir, "src", relativePath);
+			var target = Path.Combine (InstallDir, "src", (InstallDir.Contains("Xamarin.iOS")?"Xamarin.iOS":"Xamarin.Mac"), relativePath);
```
Comment (rolfbjarne) [inline tools/install-source/OpenTKSourceMangler.cs]: Coding style: space before opening parenthesis, and between between parts of expressions.

File: tools/install-source/PathManglerFactory.cs [hunk 38]
```diff
@@ -50,22 +50,37 @@ public class PathManglerFactory
 
 		public bool IsMonoPath (string path)
 		{
-			return path.StartsWith(MonoSourcePath, StringComparison.Ordinal) || !path.StartsWith(XamarinSourcePath, StringComparison.Ordinal);
+			if (path.StartsWith(XamarinSourcePath, StringComparison.Ordinal))
```
Comment (rolfbjarne) [inline tools/install-source/PathManglerFactory.cs]: Same

File: tools/install-source/PathManglerFactory.cs [hunk 39]
```diff
@@ -50,22 +50,37 @@ public class PathManglerFactory
...
+				return false;
+			var xamarinRuntimePath = XamarinSourcePath.Replace ("/src/", "/runtime/");
+			if (path.StartsWith(xamarinRuntimePath, StringComparison.Ordinal))
```
Comment (rolfbjarne) [inline tools/install-source/PathManglerFactory.cs]: Same

File: tools/install-source/PathManglerFactory.cs [hunk 40]
```diff
@@ -50,22 +50,37 @@ public class PathManglerFactory
...
+				return false;
+			return path.StartsWith(MonoSourcePath, StringComparison.Ordinal);
```
Comment (rolfbjarne) [inline tools/install-source/PathManglerFactory.cs]: Same

File: tools/install-source/PathManglerFactory.cs [hunk 41]
```diff
@@ -50,22 +50,37 @@ public class PathManglerFactory
...
+		{
+			return path.Contains ("/mcs/mcs/");
+		}
+		
 		public IPathMangler GetMangler (string path)
 		{
-			if (IsMonoPath(path)) {
-				return MonoPathMangler;
-			}
+			if (IsIgnored(path))
```
Comment (rolfbjarne) [inline tools/install-source/PathManglerFactory.cs]: Same

File: tools/install-source/PathManglerFactory.cs [hunk 42]
```diff
@@ -50,22 +50,37 @@ public class PathManglerFactory
...
+				return null;
 
 			if (IsOpenTKPath (path)) {
 				return OpenTKSourceMangler;
 			}
+			
+			if (IsMonoPath(path)) {
```
Comment (rolfbjarne) [inline tools/install-source/PathManglerFactory.cs]: Same

**Comment by spouliot** [comment]
@mandel-macaque where/when are the new tests executed ?

**Comment by mandel-macaque** [comment]
@spouliot as it is right now, the tests are only ran manually. I can add a change in that to have a make test in that dir, should we merge and add them or do it now?

**Comment by spouliot** [comment]
They can be enabled in a separate PR.

**Comment by ivanherreragl** [comment]
This build now fails to show the framework sourcecode

File: tools/install-source/InstallSourcesTests/PathManglerFactoryTests.cs [hunk 43]
```diff
@@ -0,0 +1,13 @@
...
+{
+	[TestFixture()]
+	public class PathManglerFactoryTests
+	{
+		[Test()]
+		public void TestCase()
+		{
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply tools/install-source/InstallSourcesTests/PathManglerFactoryTests.cs]: No, is me updating the pr with the tests and changing my mind. Will delete it.

**Comment by GouriKumari** [comment]
Tested with build Version: 11.7.0.106 (Visual Studio Enterprise)
Hash: 0045a9bc and I could successfully step into solution.

File: tools/install-source/Makefile [hunk 44]
```diff
@@ -49,11 +49,11 @@ MAC_MDB_FILES = \
 
 install-source-ios: install-source.exe
 	@echo "Installing source files for Xamarin.iOS"
-	$(Q) $(SYSTEM_MONO) install-source.exe $(IOS_ASSEMBLIES) --link:$(USE_SOURCE_LINKS) --mono-path=$(abspath $(MONO_PATH)) --opentk-path=$(abspath $(OPENTK_PATH)/Source) --xamarin-path=$(abspath $(SRC_PATH)) --install-dir=$(abspath $(IOS_DESTDIR))$(abspath $(MONOTOUCH_PREFIX))
+	$(Q) $(SYSTEM_MONO) install-source.exe $(IOS_ASSEMBLIES) --link:$(USE_SOURCE_LINKS) --mono-path=$(abspath $(MONO_PATH)) --opentk-path=$(abspath $(OPENTK_PATH)/Source) --xamarin-path=$(abspath $(SRC_PATH)) --install-dir=$(IOS_FRAMEWORK_DIR)/Versions/$(IOS_PACKAGE_VERSION)
```
Comment (rolfbjarne) [inline tools/install-source/Makefile]: What's the reason for this change? It looks like it will do something to your system's XI instead of the locally built one when building normally.

Comment (mandel-macaque) [reply tools/install-source/Makefile]: is not the case. Since we have the pathmap working on the build, we need to know the installation path to mangle the paths and find the correct source. Will not touch anything in the system, is just avoiding to hardcode anything in the manglers.

**Comment by mandel-macaque** [comment]
removing tag after @GouriKumari tested it.

**Comment by mandel-macaque** [comment]
Exact same failure as with https://github.com/xamarin/xamarin-macios/pull/2779 which means is probably not related with the PR.
